File: backend/core/agent_heartbeat.py
"""Agent-10 slot heartbeat — SupremeAI announces itself as alive (issue #1402).

The dashboard (`/api/agents` on the Z.ai preview) distinguishes:

    🟢 online   → heartbeat written within 90s
    🔵 assigned → slot active in AGENT_SLOT_REGISTRY.yaml, no live heartbeat
    ⚪ standby  → slot inactive

Before this loop, agent-10 always showed 🔵 "no heartbeat (tool closed?)" even
while the backend was running, because only the Z.ai dashboard (agent-11)
pinged. This module closes that gap for the self-development slot.

Storage contract (shared with the MCP tower `agent_heartbeat` tool, the
dashboard API route, and scripts/agents/heartbeat_ping.*):

    Key:   supremeai:agent-heartbeat:agent-10
    Value: JSON {slot, agentId, source, updatedAtMs, updatedAt}
    TTL:   300s (a missed ping does not immediately drop the slot)

Cadence: one SET every 45s ≈ 1,920 commands/day (≈0.4% of the Upstash
free-tier budget — quota-safe, R2-03 pattern).

Multi-account failover: the canonical Upstash PRIMARY repeatedly hits its
500k/day ceiling (see issue #1402 "Redis" note). Each tick walks the account
chain until one account accepts the write — REDIS_URL (TCP) first, then the
UPSTASH_REDIS_{,SECONDARY,TERTIARY,QUATERNARY,QUINARY}_REST_* accounts when
present in the process env. Fails soft: errors are logged and the loop
retries on the next tick — a heartbeat outage must never destabilize the
backend.

Env:
    ENABLE_AGENT_HEARTBEAT   "false" disables the loop (default: on —
                             zero-LLM-cost, honest-feature default per
                             LEARNING_LOOP/SYNAPTIC_DREAM precedent)
    AGENT_HEARTBEAT_INTERVAL seconds between pings (default 45)
"""

import asyncio
import json
import os
import time
from typing import Any

from core.logging_config import logger

SLOT = "agent-10"
AGENT_ID = "SupremeAI"
SOURCE = "backend"
DEFAULT_INTERVAL_SECONDS = 45
TTL_SECONDS = 300
KEY_PREFIX = "supremeai:agent-heartbeat:"

# Same placeholder guard as core.queue.task_queue (R2-03): booting against a
# template URL burns quota with zero useful work.
_PLACEHOLDER_TOKENS = ("<your-redis-url>", "<your", "example.com")

# Canonical account order — mirrors settings.upstash_redis_rest_pool and the
# tower/dashboard chain labels.
_ACCOUNT_LABELS = ("primary", "secondary", "tertiary", "quaternary", "quinary")


def _redis_url() -> str:
    """Resolve the real REDIS_URL (env first, settings fallback)."""
    url = os.environ.get("REDIS_URL", "")
    if not url:
        try:
            from core.config import settings

            url = getattr(settings, "redis_url", "") or ""
        except Exception:  # pragma: no cover — settings import must never break heartbeat
            url = ""
    return (url or "").strip()


def _tcp_configured() -> bool:
    url = _redis_url()
    if not url:
        return False
    return not any(token in url for token in _PLACEHOLDER_TOKENS)


def _rest_pool() -> list[tuple[str, str, str]]:
    """(label, url, token) for every configured Upstash REST account.

    Delegates to settings.upstash_redis_rest_pool, which resolves each key
    12-factor style: process env FIRST, then the Infisical vault — so the
    chain works even when the deploy env only carries a subset of the vars.
    """
    try:
        from core.config import settings

        pool = settings.upstash_redis_rest_pool
    except Exception as exc:  # pragma: no cover — settings must never break heartbeat
        logger.warning(f"⚠️ agent-10 heartbeat could not read the Upstash pool: {exc}")
        return []
    labeled: list[tuple[str, str, str]] = []
    for index, (url, token) in enumerate(pool):
        label = _ACCOUNT_LABELS[index] if index < len(_ACCOUNT_LABELS) else f"pool[{index}]"
        labeled.append((label, url, token))
    return labeled


def redis_configured_for_heartbeat() -> bool:
    """True when at least ONE account in the chain is configured."""
    if _tcp_configured():
        return True
    return bool(_rest_pool())


def _build_payload() -> str:
    now_ms = int(time.time() * 1000)
    return json.dumps(
        {
            "slot": SLOT,
            "agentId": AGENT_ID,
            "source": SOURCE,
            "updatedAtMs": now_ms,
            "updatedAt": time.strftime("%Y-%m-%dT%H:%M:%S.000Z", time.gmtime(now_ms / 1000)),
        }
    )


async def _ping_tcp(tcp_url: str, payload: str) -> str:
    import redis.asyncio as aioredis

    client = aioredis.from_url(tcp_url, decode_responses=True)
    try:
        await client.set(f"{KEY_PREFIX}{SLOT}", payload, ex=TTL_SECONDS)
    finally:
        await client.aclose()
    return "primary/tcp"


async def _ping_rest(label: str, url: str, token: str, payload: str) -> str:
    import httpx

    async with httpx.AsyncClient(timeout=10) as client:
        res = await client.post(
            url,
            json=["SET", f"{KEY_PREFIX}{SLOT}", payload, "EX", TTL_SECONDS],
            headers={"Authorization": f"Bearer {token}"},
        )
    if res.status_code != 200:
        raise RuntimeError(f"upstash[{label}] HTTP {res.status_code}: {res.text[:120]}")
    data: dict[str, Any] = res.json()
    if data.get("error"):
        raise RuntimeError(f"upstash[{label}]: {data['error']}")
    return f"{label}/rest"


async def run_agent_heartbeat_loop() -> None:
    """Long-lived heartbeat loop for slot agent-10.

    Started via AgentSupervisor in core.startup.agents. Every tick walks the
    account chain (TCP primary → REST chain) until one write succeeds.
    """
    interval = max(15, int(os.getenv("AGENT_HEARTBEAT_INTERVAL", str(DEFAULT_INTERVAL_SECONDS))))
    logger.info(
        f"✅ Agent-10 heartbeat loop started (slot={SLOT}, interval={interval}s, "
        f"ttl={TTL_SECONDS}s, key={KEY_PREFIX}{SLOT}, multi-account failover on)."
    )

    while True:
        payload = _build_payload()
        wrote_via: str | None = None
        try:
            if _tcp_configured():
                try:
                    wrote_via = await _ping_tcp(_redis_url(), payload)
                except asyncio.CancelledError:
                    raise
                except Exception as exc:
                    logger.warning(f"⚠️ agent-10 heartbeat TCP write failed: {exc}")

            if wrote_via is None:
                for label, url, token in _rest_pool():
                    try:
                        wrote_via = await _ping_rest(label, url, token, payload)
                        break
                    except asyncio.CancelledError:
                        raise
                    except Exception as exc:
                        logger.warning(f"⚠️ agent-10 heartbeat REST write failed: {exc}")

            if wrote_via is None:
                logger.warning(
                    "⚠️ agent-10 heartbeat: every configured Redis account rejected "
                    "the write this tick (non-fatal, retrying next tick)."
                )
        except asyncio.CancelledError:
            logger.info("ℹ️ Agent-10 heartbeat loop cancelled — shutting down cleanly.")
            raise
        except Exception as exc:  # pragma: no cover — absolute fail-soft guarantee
            logger.warning(f"⚠️ agent-10 heartbeat tick failed (non-fatal): {exc}")

        await asyncio.sleep(interval)

File: infrastructure/mcp-control-plane/src/registry/agent_heartbeat.ts
/**
 * Agent slot heartbeat registry (issue #1402).
 *
 * Real-time online-status layer on top of the policy-side
 * `docs/master_docs/AGENT_SLOT_REGISTRY.yaml`:
 *
 *   ┌──────────┬──────────────────────────────────────────────────────────┐
 *   │ State    │ Meaning                                                  │
 *   ├──────────┼──────────────────────────────────────────────────────────┤
 *   │ online   │ heartbeat written within ONLINE_THRESHOLD_SECONDS (90s)  │
 *   │ stale    │ heartbeat older than 90s but key not yet expired (300s)  │
 *   │ expired  │ key TTL elapsed → no runtime signal (dashboard falls     │
 *   │          │ back to the YAML policy state: assigned / standby)       │
 *   └──────────┴──────────────────────────────────────────────────────────┘
 *
 * Storage contract (shared by all writers — MCP tower, backend agent-10 loop,
 * dashboard API route, scripts/agents/heartbeat_ping.*):
 *
 *   Key:   supremeai:agent-heartbeat:<slot>          (slot = "agent-N")
 *   Value: JSON { slot, agentId, source, clientId?, updatedAtMs, updatedAt }
 *   TTL:   300s (a missed ping does not immediately drop a tool to offline)
 *
 * Writers embed the epoch ms in the value so readers can derive `age`
 * without a separate timestamp key (Upstash REST has no write-time command).
 */

import { env } from "../lib/env.js";
import Redis from "ioredis";

export const HEARTBEAT_KEY_PREFIX = "supremeai:agent-heartbeat:";
export const HEARTBEAT_TTL_SECONDS = 300;
export const HEARTBEAT_ONLINE_THRESHOLD_SECONDS = 90;
export const SLOT_PATTERN = /^agent-\d+$/;

export type HeartbeatSource = "mcp-tower" | "backend" | "dashboard" | "cli";

export type HeartbeatRecord = {
  slot: string;
  agentId: string;
  source: HeartbeatSource;
  clientId?: string;
  updatedAtMs: number;
  updatedAt: string; // ISO-8601
};

export type HeartbeatState = "online" | "stale" | "expired";

export type HeartbeatStatusEntry = HeartbeatRecord & {
  ageSeconds: number;
  state: HeartbeatState;
};

/** Validates a slot id. Returns an error message, or null when valid. */
export function validateSlot(slot: string): string | null {
  if (!slot || typeof slot !== "string") return "slot is required";
  if (!SLOT_PATTERN.test(slot)) {
    return `slot must match /^agent-\\d+$/ (e.g. "agent-4"); got "${slot}"`;
  }
  return null;
}

export function heartbeatKey(slot: string): string {
  return `${HEARTBEAT_KEY_PREFIX}${slot}`;
}

/** Derives the real-time state from a record's age (ms-precision). */
export function deriveState(
  updatedAtMs: number,
  nowMs: number = Date.now(),
): HeartbeatState {
  const ageMs = Math.max(0, nowMs - updatedAtMs);
  if (ageMs < HEARTBEAT_ONLINE_THRESHOLD_SECONDS * 1000) return "online";
  if (ageMs <= HEARTBEAT_TTL_SECONDS * 1000) return "stale";
  return "expired";
}

/** Whole-second age used for display in listings. */
export function ageSecondsOf(updatedAtMs: number, nowMs: number): number {
  return Math.max(0, Math.floor((nowMs - updatedAtMs) / 1000));
}

// ── Redis transport: Upstash multi-account fallback chain ──────────────────
// The fleet's canonical Upstash PRIMARY hit its 500k/day command ceiling
// repeatedly (issue #1402 "Redis" note; live incident during development).
// Heartbeats are liveness data — they MUST survive a quota-dead account, so
// writers walk the vault-provisioned account chain (primary → secondary →
// tertiary → quaternary → quinary; all injected by infisical_bootstrap) and
// readers merge across accounts, keeping the freshest record per slot.
// Kept separate from adapters/redis so the read-only health adapter keeps
// its narrow surface.

type HeartbeatRedisClient = {
  mode: "upstash-rest" | "ioredis-tcp";
  account: string;
  get: (key: string) => Promise<string | null>;
  setEx: (key: string, ttlSeconds: number, value: string) => Promise<void>;
  keys: (pattern: string) => Promise<string[]>;
};

type RedisAccountConfig = {
  label: string;
  restUrl?: string;
  restToken?: string;
  tcpUrl?: string;
};

const CHAIN_ENV_SUFFIXES: Array<[string, string, string]> = [
  ["secondary", "UPSTASH_REDIS_SECONDARY_REST_URL", "UPSTASH_REDIS_SECONDARY_REST_TOKEN"],
  ["tertiary", "UPSTASH_REDIS_TERTIARY_REST_URL", "UPSTASH_REDIS_TERTIARY_REST_TOKEN"],
  ["quaternary", "UPSTASH_REDIS_QUATERNARY_REST_URL", "UPSTASH_REDIS_QUATERNARY_REST_TOKEN"],
  ["quinary", "UPSTASH_REDIS_QUINARY_REST_URL", "UPSTASH_REDIS_QUINARY_REST_TOKEN"],
];

/** Ordered account chain from env (labels match the dashboard's chain). */
export function buildAccountChain(): RedisAccountConfig[] {
  const chain: RedisAccountConfig[] = [];
  const restUrl = env.redis.restUrl;
  const restToken = env.redis.restToken;
  const tcpUrl = env.redis.url;
  if (restUrl && restToken) {
    chain.push({ label: process.env.REDIS_ACCOUNT_LABEL || "primary", restUrl, restToken });
  } else if (tcpUrl) {
    chain.push({ label: process.env.REDIS_ACCOUNT_LABEL || "primary", tcpUrl });
  }
  for (const [label, urlKey, tokenKey] of CHAIN_ENV_SUFFIXES) {
    const url = process.env[urlKey];
    const token = process.env[tokenKey];
    if (url && token) chain.push({ label, restUrl: url, restToken: token });
  }
  return chain;
}

function makeRestClient(account: RedisAccountConfig): HeartbeatRedisClient {
  const restUrl = account.restUrl as string;
  const auth = `Bearer ${account.restToken}`;
  const call = async (body: unknown[]): Promise<unknown> => {
    const res = await fetch(restUrl, {
      method: "POST",
      headers: { Authorization: auth, "Content-Type": "application/json" },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(10_000),
    });
    if (!res.ok) {
      throw new Error(`Upstash REST [${account.label}] returned HTTP ${res.status}`);
    }
    const payload = (await res.json()) as { result?: unknown; error?: string };
    if (payload.error) throw new Error(`Upstash error [${account.label}]: ${payload.error}`);
    return payload.result;
  };
  return {
    mode: "upstash-rest",
    account: account.label,
    get: async (key) => (await call(["GET", key])) as string | null,
    setEx: async (key, ttl, value) => {
      await call(["SET", key, value, "EX", String(ttl)]);
    },
    keys: async (pattern) => (await call(["KEYS", pattern])) as string[],
  };
}

function makeTcpClient(account: RedisAccountConfig): HeartbeatRedisClient {
  const connect = () =>
    new (Redis as unknown as new (
      url: string,
      opts?: object,
    ) => {
      get: (key: string) => Promise<string | null>;
      set: (key: string, value: string, mode: "EX", ttl: number) => Promise<void>;
      keys: (pattern: string) => Promise<string[]>;
      disconnect: () => void;
    })(account.tcpUrl as string, { maxRetriesPerRequest: 1, connectTimeout: 5000 });
  return {
    mode: "ioredis-tcp",
    account: account.label,
    get: async (key) => {
      const c = connect();
      try {
        return await c.get(key);
      } finally {
        c.disconnect();
      }
    },
    setEx: async (key, ttl, value) => {
      const c = connect();
      try {
        await c.set(key, value, "EX", ttl);
      } finally {
        c.disconnect();
      }
    },
    keys: async (pattern) => {
      const c = connect();
      try {
        return await c.keys(pattern);
      } finally {
        c.disconnect();
      }
    },
  };
}

function makeClient(account: RedisAccountConfig): HeartbeatRedisClient {
  return account.restUrl && account.restToken
    ? makeRestClient(account)
    : makeTcpClient(account);
}

// ── Public API ─────────────────────────────────────────────────────────────

export type RecordHeartbeatInput = {
  slot: string;
  agentId?: string;
  source: HeartbeatSource;
  clientId?: string;
};

export type RecordHeartbeatResult = {
  ok: true;
  slot: string;
  agentId: string;
  heartbeatAt: string;
  ttlSeconds: number;
  redisAccount: string;
  onlineThreshold: string;
};

/**
 * Writes a heartbeat for `slot`, walking the account chain until one accepts
 * the write. The dashboard merges across the same chain, so any MCP-connected
 * agent becomes visible in real time without running its own HTTP client
 * (issue #1402 recommendation).
 */
export async function recordHeartbeat(
  input: RecordHeartbeatInput,
): Promise<RecordHeartbeatResult> {
  const slotError = validateSlot(input.slot);
  if (slotError) throw new Error(slotError);

  const chain = buildAccountChain();
  if (chain.length === 0) {
    throw new Error(
      "No Redis configuration found (set UPSTASH_REDIS_REST_URL + UPSTASH_REDIS_REST_TOKEN, or REDIS_URL) — agent heartbeats cannot be stored",
    );
  }

  const nowMs = Date.now();
  const record: HeartbeatRecord = {
    slot: input.slot,
    agentId: input.agentId?.trim() || input.slot,
    source: input.source,
    ...(input.clientId ? { clientId: input.clientId } : {}),
    updatedAtMs: nowMs,
    updatedAt: new Date(nowMs).toISOString(),
  };
  const value = JSON.stringify(record);

  const failures: string[] = [];
  for (const account of chain) {
    try {
      await makeClient(account).setEx(
        heartbeatKey(input.slot),
        HEARTBEAT_TTL_SECONDS,
        value,
      );
      return {
        ok: true,
        slot: record.slot,
        agentId: record.agentId,
        heartbeatAt: record.updatedAt,
        ttlSeconds: HEARTBEAT_TTL_SECONDS,
        redisAccount: account.label,
        onlineThreshold: `ping again within ${HEARTBEAT_ONLINE_THRESHOLD_SECONDS}s to stay 'online'`,
      };
    } catch (err) {
      failures.push((err as Error).message);
    }
  }
  throw new Error(
    `All ${chain.length} Redis account(s) rejected the heartbeat write: ${failures.join(" | ")}`,
  );
}

export type AccountReadStatus = {
  account: string;
  reachable: boolean;
  records: number;
  error?: string;
};

/** Lists every live heartbeat record with derived real-time state. */
export async function listHeartbeats(nowMs: number = Date.now()): Promise<{
  ok: true;
  now: string;
  ttlSeconds: number;
  onlineThresholdSeconds: number;
  accounts: AccountReadStatus[];
  slots: HeartbeatStatusEntry[];
}> {
  const chain = buildAccountChain();
  if (chain.length === 0) {
    throw new Error(
      "No Redis configuration found (set UPSTASH_REDIS_REST_URL + UPSTASH_REDIS_REST_TOKEN, or REDIS_URL) — agent heartbeats cannot be read",
    );
  }

  // slot → freshest record across accounts (a recovered primary may hold an
  // older copy of a slot that failed over to secondary — keep the freshest).
  const freshest = new Map<string, HeartbeatRecord>();
  const accounts: AccountReadStatus[] = [];

  for (const account of chain) {
    const status: AccountReadStatus = { account: account.label, reachable: false, records: 0 };
    try {
      const client = makeClient(account);
      const keys = await client.keys(`${HEARTBEAT_KEY_PREFIX}*`);
      status.reachable = true;
      for (const key of keys.sort()) {
        const raw = await client.get(key);
        if (!raw) continue;
        let record: HeartbeatRecord;
        try {
          record = JSON.parse(raw) as HeartbeatRecord;
        } catch {
          continue; // corrupt payload — skip rather than break the listing
        }
        status.records += 1;
        const slot = key.slice(HEARTBEAT_KEY_PREFIX.length);
        const existing = freshest.get(slot);
        if (!existing || record.updatedAtMs > existing.updatedAtMs) {
          freshest.set(slot, record);
        }
      }
    } catch (err) {
      status.error = (err as Error).message;
    }
    accounts.push(status);
  }

  const slots: HeartbeatStatusEntry[] = [...freshest.entries()]
    .map(([slot, record]) => ({
      ...record,
      slot,
      ageSeconds: ageSecondsOf(record.updatedAtMs, nowMs),
      state: deriveState(record.updatedAtMs, nowMs),
    }))
    .sort((a, b) => a.slot.localeCompare(b.slot));

  return {
    ok: true,
    now: new Date(nowMs).toISOString(),
    ttlSeconds: HEARTBEAT_TTL_SECONDS,
    onlineThresholdSeconds: HEARTBEAT_ONLINE_THRESHOLD_SECONDS,
    accounts,
    slots,
  };
}

